### ISSUE TITLE
Authentication/Security Concerns

### ISSUE THREAD

**Issue by rob-miller-777** [capitalone/Hygieia#1031]
We have been looking into the security in Hygieia and have some concerns with how it’s implemented.  It looks like there is a call to the API service to login, but after that, all the authorization is in javascript and based on a couple cookies.  Security can be bypassed completely with very little effort.  Is there any work being done to address this? If not, we might look into improving this.

**Comment by tabladrum** [comment]
There is - from LDAP, OAUTh etc to API token based authentication.

**Comment by tabladrum** [comment]
But won't complain if you help :-)

**Comment by rob-miller-777** [comment]
Yeah, @brennanananan has been working on integrating with ldap here.  I think the biggest issue we have found is that you can change a cookie value right now on the client and be any user you want to be.  We think we need to implement some sort of backend security to the api calls, which would probably mean utilizing spring security on the API.

**Comment by MarkRx** [comment]
We saw it too. It's on the list of things to look at.

**Comment by amitmawkin** [comment]
Currently the authentication from UI etc was there to only support personalization of dashboards.
All dashboard's are inherently in a read only mode, but can only be edit by the owner of those dashboards,
We need JWT for the API layer, and modify the auth to move from cookie mode to sending JWT tokens with each api request, when ever somebody login they will get a temp token that they need to send with each api request.

**Comment by mrpudn** [comment]
It might be nice to add GitHub OAuth as a near-term SSO goal. LDAP can be heavy.

**Comment by rob-miller-777** [comment]
Hey guys, just wanted to provide an update.  @brennanananan and I have put a lot of work into implementing security, and we are getting close to having it where we want.  From a users perspective, almost nothing has changed.  But the application is secure from the backend now.

**What we have done so far**
- Implemented JWT to provide stateless authentication/authorization
- Updated angular client to work with JWT tokens
- Made token expiration configurable
- Implemented redirect functionality to take you to the page you were viewing if you token expires and you have to login again
- Made secret key configurable to support having multiple instances of the API running
- Added Spring Security to Rest API
- Built in support for multiple forms of authentication (only standard Hygieia Auth implemented at this point, but adding others should be easy)
- Refactored current login to work with Spring Security
- Created a UserInfo Repository separate from the Authentication repository to support multiple forms of authentication.  This lets us control all application specific details of a user such as roles and themes independent of the authentication type. (Themes are still only handled by the angular app on the client, and saved in local storage, but this is something we can now save for a user if we want to)
- Made admin a "role" instead of just assigning admin privileges to the user named "admin". (Haven't implemented a way to assign new admins yet, so right now it works as before where the user named "admin" is given the role of admin)
- Added method level security to dashboard modification controller so only owners of the dashboard or admins can modify a dashboard
- Added support for having multiple owners of a dashboard (have not created anyway to assign multiple owners yet, but shouldn't be hard to do)
- Implemented new owner functionality in such a way that it won't break backwards compatibility

**What we have left**

- Clean up and refactor code
- Add and update some unit tests
- Document what we have done and how things work
- Document how to add additional forms of authentication (oauth, ldap, etc...)

Here is a link to the diff if you want to take a look at what we have so far: https://github.com/capitalone/Hygieia/compare/master...StateFarmIns:springSecurityIntegration

Let us know if you have any questions or concerns on what we have, should have a pull request coming shortly.

**Comment by rob-miller-777** [comment]
Do you guys think you should have to be authenticated to view dashboards?  Or should you be able to view dashboards without logging in?

**Comment by amitmawkin** [comment]
You should be able to share a dashboard with anybody, irrespective of having an account or not.

**Comment by MarkRx** [comment]
Not sure I agree. I think it should be configurable if the "anonymous" user has view access. If Hygieia runs enterprise wide I'm not sure I want anyone in the company being able to see it.

**Comment by tabladrum** [comment]
We started with a "100% transparent" dashboard mindset. Works for us very well. But I do realize that this may not fit in other organizations.

**Comment by rob-miller-777** [comment]
We will implemented with the "100% transparent" dashboard mindset, and it might be worth adding some functionality around private dashboards in the future.

**Comment by montapas** [comment]
Hi 
I have merged the code base with my hygieia codebase. After doing this I am not able to login to hygieia for both standard and ldap options.

**Ldap Error:**
------------------------------
**XMLHttpRequest cannot load http://lt060017:8080/api/login/ldap. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:1111' is therefore not allowed access. The response had HTTP status code 401.**

Standard option error:
--------------------------------------
**XMLHttpRequest cannot load http://lt060017:8080/api/login. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:1111' is therefore not allowed access.**

Please help me

**Comment by rob-miller-777** [comment]
Can you try going to the a dashboard directly, and seeing if you are able to retrieve data, or if you get CORS issues for those api calls as well?  I don't believe I made any changes that should have had this effect.

**Comment by montapas** [comment]
Thank you now I can able to standard login. I need to add extra header in StandardLoginRequestFilter.java 
but not able to login using LDAP. Please help me what to do? Also is it possible to SAML auth?

**Comment by rob-miller-777** [comment]
To be able to utilize LDAP, you have to have configured the API to utilize your ldap server.  Which means providing the url to the LDAP server, and the DN string.  Without both of these present, the LDAP authentication provider will not be created.  The properties can be found in the README for the api.

**Comment by montapas** [comment]
I provided ldap url in .properties file. But i could not find any method like any attemptauth method in  ldaploginrequestfilter file. Please help

**Comment by montapas** [comment]
I have downloaded the code from https://github.com/capitalone/Hygieia/issues/1031 
and my properties file is
auth.ldapServerUrl=ldap://xxx.xxx.xxx.xxx:389
auth.ldapUserDnPattern=OU=Users,OU=WCB,OU=xxxx,OU=yyyy,OU=zzz,OU=qqqq,DC=qq,DC=com

Please suggest me what is wrong with me.

**Comment by rob-miller-777** [comment]
Did you replace the username part with '{0}', as mentioned in the readme.  Looks like you are missing something like 'uid={0}'

**Comment by montapas** [comment]
You mean auth.ldapUserDnPattern='uid={0}',OU=Users,OU=WCB,OU=xxxx,OU=yyyy,OU=zzz,OU=qqqq,DC=qq,DC=com

**Comment by rob-miller-777** [comment]
Yeah, but without the surrounding ' '.

**Comment by montapas** [comment]
I have tried with this but no luck !!! please help me

**Comment by montapas** [comment]
I also added the following method in LdapLoginRequestFilter.java file. it is showing bad credential
@Override
	 protected void unsuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response, AuthenticationException failed)
	         throws IOException, ServletException {
	     super.unsuccessfulAuthentication(request, response, failed);
	     LOGGER.info("  failed.getMessage() : " +   failed );
	     LOGGER.info("  failed.getMessage() : " +   failed.getMessage());
	 }

**Comment by montapas** [comment]
Should I have to add any option when running API
java -jar D:\Office\Eclipse\wrkspc\Hygieia-master\api\target\api.jar --spring.config.location=D:\Office\Eclipse\collectors\dashboard.properties  
I am using this

**Comment by rob-miller-777** [comment]
To use LDAP, should only need the server url and dn string you already have.  Can you paste the stack trace in here so we can see the issue.

**Comment by montapas** [comment]
I am getting bad credential error.

**Comment by montapas** [comment]
Is there any way to override attemptAuthentication method in LdapLoginRequestFilter.java file. please help me I am stuck

**Comment by rob-miller-777** [comment]
The "attemptAuthentication" method in the LdapLoginRequestFilter is responsible for pulling the username and password out of the request, and creating a "UsernamePasswordAuthenticationToken" which will be handled by Spring Securities built in LDAP authentication provider.  From the issues you are having it seems like there are probably two potential issues you could be facing.  The first would be an improperly configured ldap server, most likely your DN string not being accurate.  The other possibility would be you are entering the wrong username/password combination on the login page.  I imagine both could give you the Bad Credential error.

**Comment by satishc1** [comment]
@rob-miller-777 @montapas You are probably using ActiveDirectory. Support for AD will be available as a PR soon.

**Comment by montapas** [comment]
any update on AD ?

**Comment by montapas** [comment]
Can I login in Hygieia without login prompt. I mean it will take my PC logged in user credential automatically and navigate to my dashboard page directly.
Please help me.
With Regards,
Tapas

**Comment by gokulknewt** [comment]
@Team,

Could you please suggest me?
We are implementing the LDAP integration with hygieia dashboard. In application.properties file
Can we use  ldaps or https protocol instead of ldap protocol.

Regards,
Gokul k

**Comment by rob-miller-777** [comment]
Have you tried something like `auth.ldapServerUrl=ldaps://{host}:{port}` in your application props?

**Comment by gokulknewt** [comment]
rob-miller-777- So far we didn't tried out. But Our Active directory for encrypted(Secure) purpose they are suggesting to use ldaps or https.Is it possible to either one we connect from Hygieia LDAP integration?

Thanks and Regards,
Gokul k

**Comment by rob-miller-777** [comment]
Hygieia does not support this functionality currently.

**Comment by gokulknewt** [comment]
Just for confirmation ,ldaps and https both it doesn't support?

**Comment by rob-miller-777** [comment]
Sorry, that comment was in regards to @montapas , I haven't attempted to use ldaps, so it is possible, but I don't have a way to test it.

**Comment by rob-miller-777** [comment]
Probably better to open new issues for these questions instead of reusing this one.

**Comment by gokulknewt** [comment]
I have_ created new one.Please any one suggest for this issue
https://github.com/capitalone/Hygieia/issues/1361
LDAP integration does it support ldaps or https protocol #1361

Thanks,
Gokul k

**Comment by htvan1** [comment]
finallllllllly.  This was really painful guys. I have been pulling my hair out trying to figure out how to get the auth token for API.  Thought the API set up  info was outdated as I did not see any settings panel.  The API start up guide needs to mention 2 things: auth.authenticationProviders=STANDARD needs to be set to allow a user to be created....and on set up the user: admin needs to be created immediately.  I looked through the start up guides again and I do not see anything referencing these facts.  It would really help to have them stated in bold.

**Comment by tabladrum** [comment]
@satishc1 @Megha849 can you see this?

**Comment by htvan1** [comment]
Also I am now not sure if the API guide's section on Securing APIs Basic Auth...is correct even when I have the API token from the admin screen.  I tried those lines of code to encode the api token I got. Then added the user name and auth token in as headers... still got 401 unauthorized.  I suppose that would be a different ticket, but a full run though example in the guide would do wonders.   Thanks @tabladrum for helping me sort out the tribal knowledge in the API guide.
